src/utils/__tests__/__snapshots__/options.spec.js.md
# Snapshot report for `build/utils/__tests__/options.spec.js`

The actual snapshot is saved in `options.spec.js.snap`.

Generated by [AVA](https://ava.li).

## should have default options

> Snapshot 1

    {
      absolute: false,
      blame: false,
      cache: true,
      config: 'test_config',
      debug: false,
      executionId: 'test',
      format: [
        'abap',
        'actionscript',
        'ada',
        'apacheconf',
        'apl',
        'applescript',
        'arduino',
        'arff',
        'asciidoc',
        'asm6502',
        'aspnet',
        'autohotkey',
        'autoit',
        'bash',
        'basic',
        'batch',
        'bison',
        'brainfuck',
        'bro',
        'c',
        'c-header',
        'clike',
        'clojure',
        'coffeescript',
        'cpp',
        'cpp-header',
        'crystal',
        'csharp',
        'csp',
        'css-extras',
        'css',
        'd',
        'dart',
        'diff',
        'django',
        'docker',
        'eiffel',
        'elixir',
        'elm',
        'erb',
        'erlang',
        'flow',
        'fortran',
        'fsharp',
        'gedcom',
        'gherkin',
        'git',
        'glsl',
        'go',
        'graphql',
        'groovy',
        'haml',
        'handlebars',
        'haskell',
        'haxe',
        'hpkp',
        'hsts',
        'http',
        'ichigojam',
        'icon',
        'inform7',
        'ini',
        'io',
        'j',
        'java',
        'javascript',
        'jolie',
        'json',
        'jsx',
        'julia',
        'keymap',
        'kotlin',
        'latex',
        'less',
        'liquid',
        'lisp',
        'livescript',
        'lolcode',
        'lua',
        'makefile',
        'markdown',
        'markup',
        'matlab',
        'mel',
        'mizar',
        'monkey',
        'n4js',
        'nasm',
        'nginx',
        'nim',
        'nix',
        'nsis',
        'objectivec',
        'ocaml',
        'opencl',
        'oz',
        'parigp',
        'pascal',
        'perl',
        'php',
        'plsql',
        'powershell',
        'processing',
        'prolog',
        'properties',
        'property',
        'protobuf',
        'pug',
        'puppet',
        'pure',
        'python',
        'q',
        'qore',
        'r',
        'reason',
        'renpy',
        'rest',
        'rip',
        'roboconf',
        'ruby',
        'rust',
        'sas',
        'sass',
        'scala',
        'scheme',
        'scss',
        'smalltalk',
        'smarty',
        'soy',
        'sql',
        'stylus',
        'swift',
        'tap',
        'tcl',
        'textile',
        'tsx',
        'tt2',
        'twig',
        'typescript',
        'vbnet',
        'velocity',
        'verilog',
        'vhdl',
        'vim',
        'visual-basic',
        'wasm',
        'wiki',
        'xeora',
        'xojo',
        'xquery',
        'yaml',
      ],
      formatsExts: {},
      gitignore: false,
      ignore: [],
      ignoreCase: false,
      listeners: [
        'statistic',
      ],
      maxLines: 500,
      maxSize: '30kb',
      minLines: 5,
      minTokens: 50,
      mode: 'mild',
      noSymlinks: false,
      output: './report',
      path: 'path/to/code',
      reporters: [
        'console',
        'time',
      ],
      reportersOptions: {},
      silent: false,
      threshold: undefined,
    }


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [2.0.11](https://github.com/kucherenko/jscpd/compare/v2.0.10...v2.0.11) (2019-02-27)


### Bug Fixes

* **options:** fix min tokens, lines and max lines options ([d0147b5](https://github.com/kucherenko/jscpd/commit/d0147b5)), closes [#208](https://github.com/kucherenko/jscpd/issues/208)



<a name="2.0.4"></a>
## [2.0.4](https://github.com/kucherenko/jscpd/compare/v2.0.3...v2.0.4) (2019-01-08)


### Bug Fixes

* **tokenizer:** split c/cpp and c-header/cpp-header formats, now files in headers detects separatly ([b9be29e](https://github.com/kucherenko/jscpd/commit/b9be29e)), closes [#188](https://github.com/kucherenko/jscpd/issues/188)
* **tokenizer:** split c/cpp and c/cpp headers formats ([97bbd55](https://github.com/kucherenko/jscpd/commit/97bbd55))



<a name="2.0.3"></a>
## [2.0.3](https://github.com/kucherenko/jscpd/compare/v2.0.2...v2.0.3) (2019-01-08)


### Bug Fixes

* **detector:** fix bug with detection duplicates in single file ([2f4eb48](https://github.com/kucherenko/jscpd/commit/2f4eb48)), closes [#189](https://github.com/kucherenko/jscpd/issues/189)



<a name="2.0.2"></a>
## [2.0.2](https://github.com/kucherenko/jscpd/compare/v2.0.1...v2.0.2) (2018-12-28)


### Bug Fixes

* **license:** change packages with GPL-2 linense to analog with MIT ([8d1081f](https://github.com/kucherenko/jscpd/commit/8d1081f))



<a name="2.0.1"></a>
## [2.0.1](https://github.com/kucherenko/jscpd/compare/v2.0.0...v2.0.1) (2018-12-28)


### Bug Fixes

* **threshold:** add ability to use 0 value for threshold ([a18fe71](https://github.com/kucherenko/jscpd/commit/a18fe71)), closes [#182](https://github.com/kucherenko/jscpd/issues/182)



<a name="2.0.0"></a>
# [2.0.0](https://github.com/kucherenko/jscpd/compare/v1.2.3...v2.0.0) (2018-12-28)


### Features

* **store:** Add LevelDb store for optimisation memory usage ([2f04dd8](https://github.com/kucherenko/jscpd/commit/2f04dd8)), closes [#66](https://github.com/kucherenko/jscpd/issues/66) [#184](https://github.com/kucherenko/jscpd/issues/184)


### BREAKING CHANGES

* **store:** add persistent store/db



<a name="1.2.3"></a>
## [1.2.3](https://github.com/kucherenko/jscpd/compare/v1.2.2...v1.2.3) (2018-12-27)


### Bug Fixes

* **jscpd:** fix bug with mulitiformats files ([7d631db](https://github.com/kucherenko/jscpd/commit/7d631db))



<a name="1.2.2"></a>
## [1.2.2](https://github.com/kucherenko/jscpd/compare/v1.2.1...v1.2.2) (2018-12-27)



<a name="1.2.1"></a>
## [1.2.1](https://github.com/kucherenko/jscpd/compare/v1.2.0...v1.2.1) (2018-12-23)


### Bug Fixes

* **blamer:** Fix unhandled promise rejection in blamer ([23a9b5d](https://github.com/kucherenko/jscpd/commit/23a9b5d)), closes [#185](https://github.com/kucherenko/jscpd/issues/185)
* **readme:** fix path to screenshot ([09ca3b8](https://github.com/kucherenko/jscpd/commit/09ca3b8))



<a name="1.2.0"></a>
# [1.2.0](https://github.com/kucherenko/jscpd/compare/v1.1.0...v1.2.0) (2018-12-14)


### Bug Fixes

* **html reporter:** Fix bug with empty lines in code blocks ([af99689](https://github.com/kucherenko/jscpd/commit/af99689))


### Features

* **html reporter:** add graph to report ([d051154](https://github.com/kucherenko/jscpd/commit/d051154))



<a name="1.1.0"></a>
# [1.1.0](https://github.com/kucherenko/jscpd/compare/v1.0.3...v1.1.0) (2018-12-02)


### Features

* **html,consoleFull reporters:** Add blamed lines to html and consoleFull reporters, add syntax hig ([1168bfb](https://github.com/kucherenko/jscpd/commit/1168bfb))
* **modes:** Add custom mode ([3aba4cd](https://github.com/kucherenko/jscpd/commit/3aba4cd)), closes [#172](https://github.com/kucherenko/jscpd/issues/172)



<a name="1.0.3"></a>
## [1.0.3](https://github.com/kucherenko/jscpd/compare/v1.0.2...v1.0.3) (2018-11-27)


### Bug Fixes

* **options.path:** Fix path options ([106a6d4](https://github.com/kucherenko/jscpd/commit/106a6d4)), closes [#177](https://github.com/kucherenko/jscpd/issues/177)



<a name="1.0.2"></a>
## [1.0.2](https://github.com/kucherenko/jscpd/compare/v1.0.1...v1.0.2) (2018-11-27)


### Bug Fixes

* support locally installed reporter and mode ([01ee5af](https://github.com/kucherenko/jscpd/commit/01ee5af))



<a name="1.0.1"></a>
## [1.0.1](https://github.com/kucherenko/jscpd/compare/v1.0.0...v1.0.1) (2018-11-27)


### Bug Fixes

* add support for trailling slash gitignore pattern ([0a48301](https://github.com/kucherenko/jscpd/commit/0a48301))



<a name="1.0.0"></a>
# [1.0.0](https://github.com/kucherenko/jscpd/compare/v1.0.0-rc.6...v1.0.0) (2018-11-21)


### Bug Fixes

* **travis:** add installation for jscpd-badge-reporter ([d09232f](https://github.com/kucherenko/jscpd/commit/d09232f))
* **travis:** add scripts running ([1bb099c](https://github.com/kucherenko/jscpd/commit/1bb099c))
* **travis:** fix scripts running ([fec133b](https://github.com/kucherenko/jscpd/commit/fec133b))
* **travis:** fix scripts running again ([bdccafb](https://github.com/kucherenko/jscpd/commit/bdccafb))



<a name="1.0.0-rc.6"></a>
# [1.0.0-rc.6](https://github.com/kucherenko/jscpd/compare/v1.0.0-rc.5...v1.0.0-rc.6) (2018-11-17)


### Features

* **reporter:** Add html reporter ([831447a](https://github.com/kucherenko/jscpd/commit/831447a))



<a name="1.0.0-rc.5"></a>
# [1.0.0-rc.5](https://github.com/kucherenko/jscpd/compare/v1.0.0-rc.4...v1.0.0-rc.5) (2018-11-15)



<a name="1.0.0-rc.4"></a>
# [1.0.0-rc.4](https://github.com/kucherenko/jscpd/compare/v1.0.0-rc.3...v1.0.0-rc.4) (2018-11-09)


### Features

* **cli:** Add ability to point few pathes to cli as args ([d1e6cfc](https://github.com/kucherenko/jscpd/commit/d1e6cfc))
* **readme,reporters,hooks:** Add hooks, change reporters interface, cleanup dependencies ([dc03298](https://github.com/kucherenko/jscpd/commit/dc03298))



<a name="1.0.0-rc.3"></a>

src/utils/options.ts
import { Command } from 'commander';
import { existsSync } from 'fs';
import { readJSONSync } from 'fs-extra';
import { dirname, isAbsolute, resolve } from 'path';
import { IOptions } from '..';
import { getSupportedFormats } from '../tokenizer/formats';

export function getOption(name: string, options?: IOptions): any {
  return options ? (options as any)[name] || (getDefaultOptions() as any)[name] : (getDefaultOptions() as any)[name];
}

export function prepareOptions(cli: Command): IOptions {
  let config: string = cli.config ? resolve(cli.config) : resolve('.jscpd.json');
  let storedConfig: any = {};
  let argsConfig: any;
  let packageJsonConfig: any;
  argsConfig = {
    minTokens: cli.minTokens ? Number(cli.minTokens) : undefined,
    minLines: cli.minLines ? Number(cli.minLines) : undefined,
    maxLines: cli.maxLines ? Number(cli.maxLines) : undefined,
    maxSize: cli.maxSize,
    debug: cli.debug,
    executionId: cli.executionId,
    silent: cli.silent,
    blame: cli.blame,
    cache: cli.cache,
    output: cli.output,
    xslHref: cli.xslHref,
    format: cli.format,
    formatsExts: parseFormatsExtensions(cli.formatsExts),
    list: cli.list,
    mode: cli.mode,
    absolute: cli.absolute,
    noSymlinks: cli.noSymlinks,
    ignoreCase: cli.ignoreCase,
    gitignore: cli.gitignore
  };

  if (cli.threshold !== undefined) {
    argsConfig.threshold = Number(cli.threshold);
  }

  if (cli.reporters) {
    argsConfig.reporters = cli.reporters.split(',');
  }

  if (cli.format) {
    argsConfig.format = cli.format.split(',');
  }

  if (cli.ignore) {
    argsConfig.ignore = cli.ignore.split(',');
  }

  argsConfig.path = cli.path ? [cli.path].concat(cli.args) : cli.args;

  Object.keys(argsConfig).forEach(key => {
    if (typeof argsConfig[key] === 'undefined') {
      delete argsConfig[key];
    }
  });

  if (!existsSync(config)) {
    config = '';
  } else {
    storedConfig = readJSONSync(config);
  }

  if (existsSync(process.cwd() + '/package.json')) {
    packageJsonConfig = readJSONSync(process.cwd() + '/package.json').jscpd || {};
  }

  const result: IOptions = {
    ...{ config },
    ...getDefaultOptions(),
    ...packageJsonConfig,
    ...storedConfig,
    ...argsConfig
  };

  if (result.hasOwnProperty('config') && result.config && isAbsolute(result.config) && result.path) {
    result.path = result.path.map((path: string) => resolve(dirname(config), path));
  }

  result.reporters = result.reporters || [];
  result.listeners = result.listeners || [];

  if (result.silent) {
    result.reporters = result.reporters.filter(reporter => reporter.indexOf('console') === -1).concat('silent');
  }

  if (result.threshold !== undefined) {
    result.reporters = [...result.reporters, 'threshold'];
  }

  result.reporters = [...result.reporters, 'time'];
  result.reporters = [...new Set(result.reporters)];
  return result;
}

export function getDefaultOptions(): IOptions {
  return {
    executionId: new Date().toISOString(),
    path: [process.cwd()],
    minLines: 5,
    maxLines: 500,
    maxSize: '30kb',
    minTokens: 50,
    output: './report',
    reporters: ['console', 'time'],
    listeners: ['statistic'],
    ignore: [],
    mode: 'mild',
    threshold: undefined,
    format: [...getSupportedFormats()],
    formatsExts: {},
    debug: false,
    silent: false,
    blame: false,
    cache: true,
    absolute: false,
    noSymlinks: false,
    ignoreCase: false,
    gitignore: false,
    reportersOptions: {}
  };
}

function parseFormatsExtensions(extensions: string): { [key: string]: string[] } | undefined {
  const result: { [key: string]: string[] } = {};

  if (extensions) {
    extensions.split(';').forEach((format: string) => {
      const pair = format.split(':');
      result[pair[0]] = pair[1].split(',');
    });
  } else {
    return undefined;
  }
  return result;
}
